src/client/gui/lib/vm_details/terminal.dart

```diff
@@ -1,102 +1,189 @@
+
+class TerminalNotifier
+    extends AutoDisposeFamilyNotifier<Terminal?, TerminalIdentifier> {
+  final lock = Lock();
   Isolate? isolate;
-  final scrollController = ScrollController();
-  final focusNode = FocusNode();
+  late final vmStatusProvider = vmInfoProvider(arg.vmName).select((info) {
+    return info.instanceStatus.status == Status.RUNNING;
+  });
 
   @override
-  void dispose() {
-    isolate?.kill(priority: Isolate.immediate);
-    scrollController.dispose();
-    focusNode.dispose();
-    super.dispose();
+  Terminal? build(TerminalIdentifier arg) {
+    ref.onDispose(_dispose);
+    lock.synchronized(_initShell).then((value) => state = value);
+    ref.listen(vmStatusProvider, (previous, next) {
+      if ((previous ?? false) && !next) stop();
+    });
+    return null;
   }
 
-  Future<void> initTerminal() async {
-    final vmShellsNotifier = ref.read(vmShellsProvider(widget.name).notifier);
-    final thisTerminal = terminal;
-    if (thisTerminal == null) return;
+  Future<Terminal?> _initShell() async {
+    final currentState = stateOrNull;
+    if (currentState != null) return currentState;
 
-    final sshInfo = await ref.read(grpcClientProvider).sshInfo(widget.name);
-    if (sshInfo == null) return;
+    final running = ref.read(vmStatusProvider);
+    if (!running) return null;
 
+    final grpcClient = ref.read(grpcClientProvider);
+    final sshInfo = await grpcClient.sshInfo(arg.vmName).onError((err, stack) {
+      ref
+          .notifyError((error) => 'Failed to get SSH information: $err')
+          .call(err, stack);
+      return null;
+    });
+    if (sshInfo == null) return null;
+
+    final terminal = Terminal(maxLines: 10000);
     final receiver = ReceivePort();
     final errorReceiver = ReceivePort();
     final exitReceiver = ReceivePort();
-    isolate = await Isolate.spawn(
-      sshIsolate,
-      SshShellInfo(
-        sender: receiver.sendPort,
-        width: thisTerminal.viewWidth,
-        height: thisTerminal.viewHeight,
-        sshInfo: sshInfo,
-      ),
-      onError: errorReceiver.sendPort,
-      onExit: exitReceiver.sendPort,
-      errorsAreFatal: true,
-    );
-
-    vmShellsNotifier.update((state) => state + 1);
 
     errorReceiver.listen((es) {
+      final (Object? error, String? stack) = (es[0], es[1]);
       logger.e(
-        'Error from ${widget.name} ssh isolate',
-        error: es[0],
-        stackTrace: es[1] != null ? StackTrace.fromString(es[1]) : null,
+        'Error from $arg ssh isolate',
+        error: error,
+        stackTrace: stack != null ? StackTrace.fromString(stack) : null,
       );
     });
+
     exitReceiver.listen((_) {
-      logger.d('Exited ${widget.name} ssh isolate');
+      logger.d('Exited $arg ssh isolate');
       receiver.close();
       errorReceiver.close();
       exitReceiver.close();
-      vmShellsNotifier.update((state) => max(0, state - 1));
-      if (mounted) setState(() => terminal = null);
+      stop();
     });
+
```

Comment:
Detail for later: top line in the terminal is cut in here:

![image](https://github.com/canonical/multipass/assets/6698114/225c8efe-9c68-4490-867f-4924b17322a4)

I don't remember resizing the window manually, but we should probably coerce the terminal size to the closest line boundary. Again, later.

Reply:
Chris also reported this issue some time ago, and after some investigation, I think the issue comes from the terminal library itself, so not something we could've fixed manually and quickly. But I just updated that dependency and I can't reproduce the issue anymore. Please let me know if you still encounter it.

Reply:
I don't see lines cut in `top` anymore, but I still see it just when the terminal has multiple lines:

![image](https://github.com/canonical/multipass/assets/6698114/f8b02f3d-2dfd-49f5-ad5a-39e0ae492e31)

We know the font, so we should be able to:

1) tweak the default size to align with character lines
2) coerce the size of the window/terminal pane to character lines

We'll need to leave 2 for later, but I imagine 1 would be a very quick tweak, no?

Reply:
Nevermind, we'll deal with this later.